src/metabase/models/user_key_value.clj

```diff
@@ -0,0 +1,29 @@
+(ns metabase.models.user-key-value
+  (:require
+   [metabase.util.malli :as mu]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]))
+
+(methodical/defmethod t2/table-name :model/UserKeyValue [_model] :user_key_value)
+
+(doto :model/UserKeyValue
+  (derive :metabase/model)
+  (derive :hook/timestamped?))
+
+(mu/defn insert!
+  "Upserts a KV-pair"
+  [user-id :- :int
+   k :- :string
+   v :- [:maybe :string]]
+  (t2/with-transaction [_]
```

Comment:
Since we don't use serializable transactions, this can fail due to a phantom read. Perhaps we should catch collisions and retry?

Reply:
I wasn't sure how to reliably catch collisions without catching other errors (given all the potential DBs we support for app-dbs), though I guess maybe retrying on any error would be fine too. Do we have prior art for retrying on a duplicate key exception, or do you think retrying on any error would be the way to go?